### PR TITLE
Language Modeling Datasets and Sampler

### PR DESCRIPTION
**Pull request by szha**
## Description ##
Add language modeling dataset wikitext-2 and wikitext-103. Add interval sampler that is suitable for batched language model training. Update word-language-model example.

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Add wikitext-2, wikitext-103, and tests.
- [x] Add Interval Sampler, and test.
- [x] Update example to use new dataset.

### CODE REVIEW

File: python/mxnet/gluon/data/sampler.py [hunk 1]
```diff
@@ -136,3 +136,30 @@ def __len__(self):
...
             "but got %s"%self._last_batch)
+
+
+class IntervalSampler(Sampler):
+    """Samples elements from [0, length) at fixed intervals.
+
+    Parameters
+    ----------
+    length : int
+        Length of the sequence.
```
Comment (zhreshold) [inline python/mxnet/gluon/data/sampler.py]: interval : int
    Sampling interval

File: python/mxnet/gluon/data/sampler.py [hunk 2]
```diff
@@ -136,3 +136,30 @@ def __len__(self):
...
+
+    Examples
+    --------
+    >>> sampler = gluon.data.IntervalSampler(13, interval=3)
+    >>> list(sampler)
+    [0, 3, 6, 9, 12, 1, 4, 7, 10, 2, 5, 8, 11]
+    """
+    def __init__(self, length, interval):
+        self._length = length
+        self._interval = interval
```
Comment (zhreshold) [inline python/mxnet/gluon/data/sampler.py]: add range check?

File: python/mxnet/gluon/data/text.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    Each sample is a vector of length equal to the specified sequence length.
+    At the end of each sentence, an end-of-sentence token '<eos>' is added.
+
+    Parameters
+    ----------
+    root : str, default '~/.mxnet/datasets/cifar10'
+        Path to temp folder for storing data.
+    segment : str, default 'train'
+        Dataset segment. Options are 'train', 'validation', 'test'.
+    indexer : :class:`~mxnet.contrib.text.indexer.TokenIndexer`, default None
```
Comment (piiswrong) [inline python/mxnet/gluon/data/text.py]: I wouldn't expose this to users.
1. Indexer is not the standard term for this.
2. This is contrib API and subject to change. Gluon Dataset should use a separate vocabulary API

File: python/mxnet/gluon/data/text.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+class WikiText2(dataset._DownloadedDataset):
+    """WikiText-2 word-level dataset for language modeling, from Salesforce research.
+
+    From
+    https://einstein.ai/research/the-wikitext-long-term-dependency-language-modeling-dataset
+
+    License: Creative Commons Attribution-ShareAlike
+
+    Each sample is a vector of length equal to the specified sequence length.
+    At the end of each sentence, an end-of-sentence token '<eos>' is added.
```
Comment (piiswrong) [inline python/mxnet/gluon/data/text.py]: if seq_len doesn't respect sentence boundary why should it end with eos?

File: python/mxnet/gluon/data/text.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    ----------
+    root : str, default '~/.mxnet/datasets/cifar10'
+        Path to temp folder for storing data.
+    segment : str, default 'train'
+        Dataset segment. Options are 'train', 'validation', 'test'.
+    indexer : :class:`~mxnet.contrib.text.indexer.TokenIndexer`, default None
+        The indexer to use for indexing the text dataset. If None, a default indexer is created.
+    seq_len : int, default 35
+        The sequence length of each sample, regardless of the sentence boundary.
+    transform : function, default None
```
Comment (piiswrong) [inline python/mxnet/gluon/data/text.py]: Dataset now has a transform API. Use that instead of adding transform callback to every dataset

File: python/mxnet/gluon/data/sampler.py [hunk 6]
```diff
@@ -136,3 +136,30 @@ def __len__(self):
...
+    Parameters
+    ----------
+    length : int
+        Length of the sequence.
+
+    Examples
+    --------
+    >>> sampler = gluon.data.IntervalSampler(13, interval=3)
+    >>> list(sampler)
+    [0, 3, 6, 9, 12, 1, 4, 7, 10, 2, 5, 8, 11]
```
Comment (piiswrong) [inline python/mxnet/gluon/data/sampler.py]: why should it roll over at the end?

File: python/mxnet/gluon/data/text.py (hunk 3 above)
Comment (szha) [reply python/mxnet/gluon/data/text.py]: Thanks.
1. I think it's safe to say that indexer is a clear enough term to reflect what it does.
2. If I understand correctly, I believe the indexer class is intended to serve the same purpose as what you call 'vocabulary'

There is a reason this needs to be exposed. Suppose we have training dataset whose vocabulary is {a, b, c} plus unknown tokens, an test dataset has vocabulary {a, b, d}. As a standard practice, the token 'd' in the test dataset should be indexed as unknown. This means the indexing of test dataset depends on the index from training dataset.

File: python/mxnet/gluon/data/text.py (hunk 4 above)
Comment (szha) [reply python/mxnet/gluon/data/text.py]: Even though the sentence boundary is not considered in providing sample chunks, it's still necessary for language model to be able to predict where sentence ends. In that sense, these concepts are orthogonal.

File: python/mxnet/gluon/data/text.py (hunk 5 above)
Comment (szha) [reply python/mxnet/gluon/data/text.py]: Sure, I can take a look at that. What about vision? Are the existing transform options dropped? Where can I find relevant discussion?

File: python/mxnet/gluon/data/sampler.py (hunk 6 above)
Comment (szha) [reply python/mxnet/gluon/data/sampler.py]: Is there a reason that you think it shouldn't? I think this sampler should exhaust every sample in a dataset. If for some reason it needs to drop some samples, for the purpose of mini-batch for example, then a wrapper sampler should take care of that.

File: python/mxnet/gluon/data/text.py (hunk 3 above)
Comment (piiswrong) [reply python/mxnet/gluon/data/text.py]: 1. I don't think so. Do you have reference of it being used somewhere?
2. It is, but it is contrib API. If you want to use it directly then gluon.data.text need to be in gluon.contrib too.

We do need to expose something like this. But it can't be TokenIndexer.

**Comment by piiswrong** [comment]
This Dataset.indexer design doesn't work for the use case where you want to combine (or take intersection of) the vocab of too datasets (like train and val)

**Comment by szha** [comment]
Indeed, I wasn't considering such case because it isn't good practice to index using anything other than training set. That said, providing an option to update the input indexer should be sufficient to cover this case. Would that be OK?

File: python/mxnet/gluon/data/text.py (hunk 3 above)
Comment (szha) [reply python/mxnet/gluon/data/text.py]: Do you have something else in mind?

Also, what should I provide here in place of TokenIndexer? If you could help me understand the reasoning for "it can't be TokenIndexer", I can help propose alternatives too.

Comment (szha) [reply python/mxnet/gluon/data/text.py]: I think it's probably a safer bet to move the dataset to contrib first.

File: python/mxnet/gluon/data/sampler.py (hunk 1 above)
Comment (szha) [reply python/mxnet/gluon/data/sampler.py]: Added docstring.

File: python/mxnet/gluon/data/sampler.py (hunk 6 above)
Comment (piiswrong) [reply python/mxnet/gluon/data/sampler.py]: The name interval sampler suggests it should behave like [begin:end:step]

Comment (szha) [reply python/mxnet/gluon/data/sampler.py]: I see the confusion now. Should I add an option to specify whether to roll over?

Comment (piiswrong) [reply python/mxnet/gluon/data/sampler.py]: This doesn't seem very generic anyway. I would put it in examples

**Comment by piiswrong** [comment]
Then you would have problem when you want only the top 2000 tokens.

Since this is not a very common use case, I think the current version is fine for contrib

File: python/mxnet/gluon/data/sampler.py (hunk 6 above)
Comment (szha) [reply python/mxnet/gluon/data/sampler.py]: This sampler is needed for any long-form text processing that requires passing hidden state from sample to sample. I'd expect repeated use for this, which is why I chose to put it here. Do you prefer to update its interface for handling roll-over, or to move this contrib, or do you still prefer it to be dropped?

**Comment by zhreshold** [comment]
I am more concerned with the name `Indexer` and the related behavior(manually extracting Indexer and synchronize between train/valid dataset). The others LGTM now.

**Comment by szha** [comment]
@zhreshold thanks. @astonzhang shall we consider a different name for indexer, like the aforementioned "vocabulary"?

File: python/mxnet/gluon/data/sampler.py (hunk 6 above)
Comment (szha) [reply python/mxnet/gluon/data/sampler.py]: @piiswrong ping

**Comment by astonzhang** [comment]
I would like to propose the following change for class names:
TokenIndexer -> Vocabulary
Glossary -> VocabularyEmbedding
TokenEmbedding -> PretrainedEmbedding

Otherwise, having both Vocabulary and Glossary is likely confusing. Having both VocabularyEmbedding and TokenEmbedding is also likely confusing. 

Is the proposed change OK?

**Comment by szha** [comment]
VocabularyEmbedding and PretrainedEmbedding don't sound like they would inherit each other, and the concerns are unclear just based on the name. I probably won't remember which is which after a couple of weeks. Let's consider other names for those two.

**Comment by astonzhang** [comment]
How about:
TokenIndexer -> Vocabulary
Glossary -> VocabularyEmbedding
TokenEmbedding -> Embedding

**Comment by tolkien16** [comment]
Tomasz

**Comment by tolkien16** [comment]
MONSTER

**Comment by tolkien16** [comment]
MONSTER 2

**Comment by tolkien16** [comment]
MONSTER

**Comment by szha** [comment]
CompositeEmbedding sounds more like what Glossary does.

**Comment by szha** [comment]
I updated this PR based on the latest change in text api naming. Also, I made the vocabulary as a property of the dataset for exchanging the index. Feel free to comment and I'd like to get this merged once 1.1 release is cut.

**Comment by szha** [comment]
To address the concern of merging datasets based on frequencies, I made the frequencies counts a property of the dataset too. This way, user has the control on how vocabulary is made.

Currently the tokenization is naive and the next step should be to have a proper tokenizer class. Once that's available, the datasets should expose an option for specifying tokenizers.

**Comment by szha** [comment]
@zhreshold @piiswrong pinging for another pass of review.

**Comment by szha** [comment]
Thanks. I will wait another day before merging, in case @piiswrong has additional feedback.

**Comment by szha** [comment]
Connected offline with @piiswrong that current design is OK to check in in contrib package.
